File: leetcode/src/anagrams.cpp
/*
 * given an array of strings, return all groups of strings that are anagrams. All inputs are in lower-case
 * ps, anagram means words which have same characters with different order, e.g dog and god
 * */
#include "../include/preliminary.h"
#include <set>
#include <map>

typedef multimap<string, int>::iterator  iterType;
typedef pair<string, int>  pairType;

class Solution{

public:
    vector<string> anagrams(vector<string>& strs){
        vector<string> res;
        if(strs.empty())    return res;

        const int n = strs.size();
        multimap<string, int> mmp;  //pair(key of word, index in strs)
        set<string>    keys;        //save key of word
        for(int i = 0;i < n; i++){
            string kstr = wordkey(strs[i]);
            keys.insert(kstr);
            mmp.insert(pairType(kstr, i));
        }
    
        for(set<string>::iterator kIter = keys.begin(); kIter != keys.end(); ++kIter){

            if(mmp.count(*kIter) == 1)    continue;

            pair<iterType, iterType> ret = mmp.equal_range(*kIter);   //(lowerbound, upperbound) multimap::equal_range()

            for(iterType vIter = ret.first; vIter != ret.second; ++vIter){
                res.push_back( strs[vIter->second] );
            }
        }
        return res;
    }

    string wordkey(const string& word){ // count sort word, as its key
        char A[26];
        memset(A, 0, sizeof(A));

        const int n = word.size();
        for(int i = 0; i < n; i++){
            A[word[i] - 'a']++;
        }

        for(int i = 1;i < 26; i++){
            A[i] += A[i-1];
        }

        char B[n+1];
        memset(B, 0, sizeof(B));

        for(int i = 0; i < n; i++){
            int chIndex = word[i] - 'a';
            B[ A[chIndex] - 1 ] = word[i];
            A[chIndex]--;
        }
        B[n] = '\0';

        return string(B);
    }
};

File: leetcode/java/anagrams.java
/*
 * given an array of strings, return all groups of strings that are anagrams. e.g dog and god are anagrams
 * test case:
 * input:dog, god, cat, tac, abc
 * output: dog, god, cat, tac
 * */
import java.io.*;
import java.util.*;

public class anagrams{
    private String wordkey(String word){  // count sort
        int[] A = new int[26];
        int n = word.length();
        for(int i=0; i < n; i++){
            A[word.charAt(i) - 'a']++;
        }
        for(int i=1; i < 26; i++){
            A[i] += A[i-1];
        }
        char[] B = new char[n];
        for(int i=0; i < n; i++){
            int idxChar = word.charAt(i) - 'a';
            B[A[idxChar] - 1] = word.charAt(i);
            A[idxChar]--;
        }
        return new String(B);
    }

    public ArrayList<String> getAnagrams(String[] strs){
        ArrayList<String> res = new ArrayList<String>();
        int n = strs.length;
        if(n < 2)    return res;
        HashMap<String, List<Integer>> mmp
            = new HashMap<String, List<Integer>>();  //replacement of multimap<>
        for(int i=0; i < n; i++){
            String kstr = wordkey(strs[i]);
            if(!mmp.containsKey(kstr)){
                mmp.put(kstr, new ArrayList<Integer>());
            }
            mmp.get(kstr).add(i);
        }
        
        for(String k : mmp.keySet()){  // Set<KeyType> 
            if(mmp.get(k).size() < 2)    continue;
            for(Integer i : mmp.get(k)){
                res.add(strs[i]);
            }
        }
        return res;
    }

    /* unit test code is in ../java_unittest/anagrams_unittest/  */
    
    /*
    private void showList(List<String> lst){
        for(String s : lst){
            System.out.print(s + " ");
        }
        System.out.println();
    }

    public void test_01(){
        Scanner scan = new Scanner(System.in);
        while(true){
            System.out.println("please input String[]");
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            StringTokenizer t = new StringTokenizer(str, " ,");
            int n = t.countTokens();
            String[] strs = new String[n];
            for(int i=0;i<n;i++){
                strs[i] = t.nextToken();
            }
            ArrayList<String> res = anagrams(strs);
            showList(res);
        }
    }
    */
}
